### PR TITLE
Support CP layout conversion in hybrid models

### PR DESCRIPTION
**Pull request by Mellonta**
## Background

TE softmax attention requires a zigzag CP layout, while linear attention does not. With chunkwise CP, using the zigzag layout for linear-attention layers doubles their communication. This PR implements the subset of #6515 needed for Nemotron, with additional optimizations. It adds the standalone `megatron/core/context_parallel` module only, nothing calls it yet. The integration is in a separate PR #6776.

## Summary
- Add contiguous ↔ zigzag redistribution for dense and packed sequences, with per-sequence padding to the `2 × CP` chunk granularity that TE requires for attention.
- Use a single direct TP × CP all-to-all-v with sequence parallelism, instead of separate TP and CP exchanges.
- Add `ContextParallelLayoutManager` to drive the per-layer transitions from a layer-layout list, so callers only supply the layout policy (it will be used in integration PR).

## Tests
```
uv run python -m torch.distributed.run --nproc-per-node 4 -m pytest -q \
    tests/unit_tests/context_parallel/test_layout.py
```

### CODE REVIEW

**Review by deepakn94 — APPROVED** [review]

File: megatron/core/models/hybrid/hybrid_block.py [hunk 1]
```diff
@@ -35,6 +42,9 @@
 from megatron.core.transformer.utils import sharded_state_dict_default
 from megatron.core.utils import WrappedTensor, deprecate_inference_params, make_viewless_tensor
 
+CPLayout = Literal["contiguous", "zigzag"]
```
Comment (Phlip79) [inline megatron/core/models/hybrid/hybrid_block.py]: Defining `CPLayout` does not belong in this file. You already have this defined in `context_parallel_layout.py` as `_CPLayout`.

File: megatron/core/models/hybrid/hybrid_block.py [hunk 2]
```diff
@@ -35,6 +42,9 @@
...
+_ATTENTION_LAYER_TYPES = (LayerSymbols.ATTENTION, LayerSymbols.DS_ATTENTION, LayerSymbols.MLA)
```
Comment (Phlip79) [inline megatron/core/models/hybrid/hybrid_block.py]: How about you define this within `LayerSymbols`? Similar to `VALID_LAYERS` defined [here](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/models/hybrid/hybrid_layer_allocation.py#L26C5-L26C17).

File: megatron/core/context_parallel_layout.py [hunk 3]
```diff
@@ -0,0 +1,670 @@
...
+        output = received.new_zeros((plan.zigzag_local_token_count, *received.shape[1:]))
+        output.index_copy_(0, plan.forward_receive_positions, received)
+        return output
+    return received.index_select(0, plan.reverse_receive_indices)
+
+
+def contiguous_to_zigzag(
+    input_: torch.Tensor,
+    cp_group: torch.distributed.ProcessGroup,
+    *,
```
Comment (Phlip79) [inline megatron/core/context_parallel_layout.py]: Why is this not explicit? You shouldn't need to use `*`.

File: megatron/core/models/hybrid/hybrid_block.py [hunk 4]
```diff
@@ -270,6 +295,108 @@ def mamba_state_shapes_per_request(self) -> Optional[Tuple[Tuple[int], Tuple[int
...
+    ) -> Tensor:
+        """Convert between CP layouts when a hybrid layer boundary requires it."""
+        if not self._requires_attention_layout_conversion or source_layout == target_layout:
+            return hidden_states
+        layout_kwargs = {
+            "sequence_parallel": self.config.sequence_parallel,
+            "tp_group": self.tp_group,
+            "tp_cp_group": self.tp_cp_group,
+            "thd_plan": thd_plan,
+        }
```
Comment (Phlip79) [inline megatron/core/models/hybrid/hybrid_block.py]: Why is this a dict?

File: megatron/core/models/hybrid/hybrid_block.py [hunk 5]
```diff
@@ -270,6 +295,108 @@ def mamba_state_shapes_per_request(self) -> Optional[Tuple[Tuple[int], Tuple[int
                 return layer.self_attention.mamba_state_shapes_per_request()
         return None
 
+    def _get_layer_cp_layout(self, layer_index: int) -> CPLayout:
+        """Return the CP layout used to execute one hybrid layer."""
+        if self.layer_type_list[layer_index] in _ATTENTION_LAYER_TYPES:
+            return self.config.attention_cp_layout
+        return self.config.linear_cp_layout
+
+    def _convert_cp_layout(
```
Comment (Phlip79) [inline megatron/core/models/hybrid/hybrid_block.py]: Why not have this function within `context_parallel_layout.py`?

File: megatron/core/models/hybrid/hybrid_block.py [hunk 6]
```diff
@@ -270,6 +295,108 @@ def mamba_state_shapes_per_request(self) -> Optional[Tuple[Tuple[int], Tuple[int
...
+            "tp_cp_group": self.tp_cp_group,
+            "thd_plan": thd_plan,
+        }
+        if source_layout == "contiguous" and target_layout == "zigzag":
+            return contiguous_to_zigzag(hidden_states, self.cp_group, **layout_kwargs)
+        if source_layout == "zigzag" and target_layout == "contiguous":
+            return zigzag_to_contiguous(hidden_states, self.cp_group, **layout_kwargs)
+        raise ValueError(f"Unsupported CP layout conversion: {source_layout} to {target_layout}")
+
+    def _prepare_layer_input(
```
Comment (Phlip79) [inline megatron/core/models/hybrid/hybrid_block.py]: Same here. Why is this defined in this file?

File: megatron/core/models/hybrid/hybrid_block.py [hunk 7]
```diff
@@ -270,6 +295,108 @@ def mamba_state_shapes_per_request(self) -> Optional[Tuple[Tuple[int], Tuple[int
...
+        if layer_index != len(self.layer_type_list) - 1:
+            return hidden_states
+        return self._convert_cp_layout(
+            hidden_states,
+            self._get_layer_cp_layout(layer_index),
+            self.config.linear_cp_layout,
+            thd_plan,
+        )
+
+    def _build_packed_attention_layout(
```
Comment (Phlip79) [inline megatron/core/models/hybrid/hybrid_block.py]: Same here. Why is this function in this file?

File: megatron/core/models/hybrid/hybrid_block.py [hunk 8]
```diff
@@ -369,37 +514,47 @@ def get_inner_quant_context(config, layer_number):
...
                         else:  # MambaLayer, Expert, or MLP
-                            hidden_states = layer(
+                            layer_kwargs = dict(
                                 hidden_states=hidden_states,
                                 attention_mask=attention_mask,
                                 inference_context=inference_context,
-                                packed_seq_params=packed_seq_params,
+                                packed_seq_params=layer_packed_seq_params,
                             )
+                            hidden_states = layer(**layer_kwargs)
```
Comment (Phlip79) [inline megatron/core/models/hybrid/hybrid_block.py]: Why are we now creating a `dict`?

File: megatron/core/context_parallel/layout.py
Comment (Phlip79) [inline megatron/core/context_parallel/layout.py:1]: This file does not belong in `megatron/core`. How about you create a new directory `context_parallel/`, similar to what is done with `tensor_parallel/` and `pipeline_parallel/`?

**Review by Phlip79 — CHANGES_REQUESTED** [review]
`megatron/core/models/hybrid/hybrid_block.py` should remain as minimal as possible (within reason). A lot of the logic added in this PR should be placed elsewhere.

This PR is also way too big. Can you add the logic in one PR and integrate it in another?

**Comment by zhongbozhu** [comment]
What's the plan for merging this https://github.com/NVIDIA/Megatron-LM/pull/6515? Seems like they are quite similar.

**Review by zhongbozhu — APPROVED** [review]

File: megatron/core/context_parallel/layout.py [hunk 9]
```diff
@@ -0,0 +1,825 @@
...
+        )
+
+    # CP attention needs two equal chunks per CP rank. Pad each sequence to that granularity,
+    # then add only enough padding to the last sequence to split the batch evenly over TP.
+    local_target_token_count = target_token_count // cp_size
+    tp_padding = (-local_target_token_count) % tp_size
+    if tp_padding:
+        target_lengths = target_lengths.clone()
+        target_lengths[-1] += tp_padding * cp_size
+        target_token_count += tp_padding * cp_size
```
[the remark this answers is not in the document]
Comment (Mellonta) [reply megatron/core/context_parallel/layout.py:417]: never happen, rejected in validation

**Comment by Mellonta** [comment]
Addressed reasonable comments from Claude

**Review by Phlip79 — APPROVED** [review]
